File: 20q-kakao-bot/src/main/kotlin/party/qwer/twentyq/bridge/handlers/AdminHandler.kt
package party.qwer.twentyq.bridge.handlers

import jakarta.annotation.PreDestroy
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.cancel
import kotlinx.coroutines.delay
import kotlinx.coroutines.launch
import org.slf4j.LoggerFactory
import org.springframework.stereotype.Component
import party.qwer.twentyq.config.AppProperties
import party.qwer.twentyq.redis.RiddleSessionRepository
import party.qwer.twentyq.service.RiddleService
import party.qwer.twentyq.service.riddle.MetaQuestionValidator
import party.qwer.twentyq.util.common.security.requireAdminOrThrow
import party.qwer.twentyq.util.common.security.requireSessionOrThrow
import party.qwer.twentyq.util.game.GameMessageProvider

@Component
class AdminHandler(
    private val appProperties: AppProperties,
    private val riddleService: RiddleService,
    private val sessionRepo: RiddleSessionRepository,
    private val metaQuestionValidator: MetaQuestionValidator,
    private val messageProvider: GameMessageProvider,
) {
    companion object {
        private val log = LoggerFactory.getLogger(AdminHandler::class.java)
        private const val RESTART_DELAY_MS = 3000L
    }

    private val adminScope = CoroutineScope(SupervisorJob() + Dispatchers.IO)

    @PreDestroy
    fun cleanup() {
        adminScope.cancel()
    }

    suspend fun forceEnd(
        chatId: String,
        userId: String,
    ): String {
        log.info("HANDLE_ADMIN_FORCE_END chatId={}, userId={}", chatId, userId)
        requireAdminOrThrow(
            adminUserIds = appProperties.admin.userIds,
            userId = userId,
            chatId = chatId,
            logger = log,
            warnMessage = "ADMIN_PERMISSION_DENIED userId={}, chatId={}",
            warnArgs = arrayOf(userId, chatId),
            errorMessage = messageProvider.get("error.no_permission"),
        )
        riddleService.requireSessionOrThrow(chatId)

        val result = riddleService.surrender(chatId)
        log.info("ADMIN_FORCE_END_SUCCESS chatId={}, adminId={}", chatId, userId)

        return messageProvider.get("admin.force_end_prefix") + result
    }

    suspend fun clearAll(
        chatId: String,
        userId: String,
    ): String {
        log.info("HANDLE_ADMIN_CLEAR_ALL chatId={}, userId={}", chatId, userId)

        requireAdminOrThrow(
            adminUserIds = appProperties.admin.userIds,
            userId = userId,
            chatId = chatId,
            logger = log,
            warnMessage = "ADMIN_PERMISSION_DENIED userId={}, chatId={}",
            warnArgs = arrayOf(userId, chatId),
            errorMessage = messageProvider.get("error.no_permission"),
        )

        sessionRepo.clearAllData(chatId)
        log.info("ADMIN_CLEAR_ALL_SUCCESS chatId={}, adminId={}", chatId, userId)

        return messageProvider.get("admin.clear_all_success")
    }

    suspend fun refreshCache(
        chatId: String,
        userId: String,
    ): String {
        log.info("HANDLE_ADMIN_REFRESH_CACHE chatId={}, userId={}", chatId, userId)
        requireAdminOrThrow(
            adminUserIds = appProperties.admin.userIds,
            userId = userId,
            chatId = chatId,
            logger = log,
            warnMessage = "ADMIN_PERMISSION_DENIED userId={}, chatId={}",
            warnArgs = arrayOf(userId, chatId),
            errorMessage = messageProvider.get("error.no_permission"),
        )

        val guardSuccess = metaQuestionValidator.refreshCache()

        log.info(
            "ADMIN_REFRESH_CACHE_SUCCESS adminId={}, guard={}",
            userId,
            guardSuccess,
        )

        return if (guardSuccess) {
            messageProvider.get("admin.cache_refresh_success")
        } else {
            messageProvider.get("admin.cache_refresh_failed")
        }
    }

    suspend fun restartAllBots(
        chatId: String,
        userId: String,
    ): String {
        log.info("HANDLE_ADMIN_RESTART_ALL chatId={}, userId={}", chatId, userId)
        requireAdminOrThrow(
            adminUserIds = appProperties.admin.userIds,
            userId = userId,
            chatId = chatId,
            logger = log,
            warnMessage = "ADMIN_PERMISSION_DENIED userId={}, chatId={}",
            warnArgs = arrayOf(userId, chatId),
            errorMessage = messageProvider.get("error.no_permission"),
        )

        // 응답 전송 후 재시작 (3초 딜레이)
        adminScope.launch {
            delay(RESTART_DELAY_MS)
            ProcessBuilder()
                .command("bash", "-c", "/home/kapu/gemini/scripts/restart_all_bots.sh")
                .start()
        }

        log.info("ADMIN_RESTART_ALL_INITIATED adminId={}", userId)
        return messageProvider.get("admin.restart_all_initiated")
    }
}

File: game-bot-go/internal/twentyq/service/admin_handler.go
package service

import (
	"context"
	"fmt"
	"log/slog"

	"github.com/park285/llm-kakao-bots/game-bot-go/internal/common/messageprovider"
	qmessages "github.com/park285/llm-kakao-bots/game-bot-go/internal/twentyq/messages"
	"github.com/park285/llm-kakao-bots/game-bot-go/internal/twentyq/redis"
)

// RiddleServiceForAdmin AdminHandler가 의존하는 RiddleService 메서드 인터페이스.
type RiddleServiceForAdmin interface {
	Surrender(ctx context.Context, chatID string) (string, error)
}

// AdminHandler 관리자 명령어 핸들러.
type AdminHandler struct {
	adminUserIDs  []string
	riddleService RiddleServiceForAdmin
	sessionStore  *redis.SessionStore
	msgProvider   *messageprovider.Provider
	logger        *slog.Logger
}

// NewAdminHandler 생성자.
func NewAdminHandler(
	adminUserIDs []string,
	riddleService RiddleServiceForAdmin,
	sessionStore *redis.SessionStore,
	msgProvider *messageprovider.Provider,
	logger *slog.Logger,
) *AdminHandler {
	return &AdminHandler{
		adminUserIDs:  adminUserIDs,
		riddleService: riddleService,
		sessionStore:  sessionStore,
		msgProvider:   msgProvider,
		logger:        logger,
	}
}

// IsAdmin 관리자 여부 확인.
func (h *AdminHandler) IsAdmin(userID string) bool {
	for _, id := range h.adminUserIDs {
		if id == userID {
			return true
		}
	}
	return false
}

// ForceEnd 관리자 강제 종료.
func (h *AdminHandler) ForceEnd(ctx context.Context, chatID string, userID string) (string, error) {
	h.logger.Info("HANDLE_ADMIN_FORCE_END", "chatID", chatID, "userID", userID)
	if !h.IsAdmin(userID) {
		h.logger.Warn("ADMIN_PERMISSION_DENIED", "userID", userID, "chatID", chatID)
		return h.msgProvider.Get(qmessages.ErrorNoPermission), nil
	}

	// 세션 확인
	session, err := h.sessionStore.Get(ctx, chatID)
	if err != nil {
		return "", fmt.Errorf("session store get: %w", err)
	}
	if session == nil {
		return h.msgProvider.Get(qmessages.ErrorNoSessionShort), nil
	}

	result, err := h.riddleService.Surrender(ctx, chatID)
	if err != nil {
		return "", fmt.Errorf("surrender: %w", err)
	}

	h.logger.Info("ADMIN_FORCE_END_SUCCESS", "chatID", chatID, "adminID", userID)
	return h.msgProvider.Get(qmessages.AdminForceEndPrefix) + result, nil
}

// ClearAll 관리자 전체 삭제.
func (h *AdminHandler) ClearAll(ctx context.Context, chatID string, userID string) (string, error) {
	h.logger.Info("HANDLE_ADMIN_CLEAR_ALL", "chatID", chatID, "userID", userID)
	if !h.IsAdmin(userID) {
		h.logger.Warn("ADMIN_PERMISSION_DENIED", "userID", userID, "chatID", chatID)
		return h.msgProvider.Get(qmessages.ErrorNoPermission), nil
	}

	if err := h.sessionStore.ClearAllData(ctx, chatID); err != nil {
		h.logger.Error("session_store_clear_all_failed", "error", err, "chatID", chatID)
		return "", fmt.Errorf("clear all data: %w", err)
	}

	h.logger.Info("ADMIN_CLEAR_ALL_SUCCESS", "chatID", chatID, "adminID", userID)
	return h.msgProvider.Get(qmessages.AdminClearAllSuccess), nil
}
